server/aesdsocket.c

```diff
@@ -0,0 +1,423 @@
+        }
+        else {     
+            d_mode = 1;
+            syslog(LOG_INFO,"aesdsocket: -d is detected \n");
+        }
+    }
+    rc = aesdsocket_server(d_mode);
+    closelog();
+    return rc;
+    
+}
+
+/*
+* Socket server API
+* 
+* Parameters:
+*   d_mode: Flag stating if daemon mode is enabled or not, 1= daemon mode
+*
+* Returns: 0 if the function executed without any error. Otherwise, error code 
+*          is logged in syslog and the program exists with code non zero value
+*   
+*/
+static int aesdsocket_server(int d_mode) {
+    int soc_server = -1;
+    int soc_client = -1;
+    int rc = -1;
+    int cmd_option = 1;
+    struct sockaddr_in aesdsoc_addr;
+    int aesdsoc_addr_len = sizeof(struct sockaddr_in);
+    int rcv_data_len = 0;
+    int wr_pointer = 0;
+    int buffer_size = FIXED_RD_BUF_SIZE;
+    FILE *fp;
+    struct sigaction signal_action;
+    char *buffer = NULL;
+    char *file_buffer = NULL;
+    char *tx_buf = NULL;
+    struct timespec start;
+    struct timespec now;
+    int diff_time = 0;
+
+
+    if(d_mode == TRUE) {
+        rc = daemon(0,0);
+        if (rc < 0) {
+            syslog(LOG_ERR, "aesdsocket: daemon creation error %s", strerror(errno));
+            goto  return_func;
+        }
+    }
+
+    fp =  fopen("/var/tmp/aesdsocketdata", "a+");
+    if (fp == NULL) {
+        syslog(LOG_ERR, "aesdsocket: File open Error %s", strerror(errno));
+        goto  return_func;
+    }
+
+    file_close = TRUE;
+
+    syslog(LOG_INFO, "aesdsocket:Starting aesdsocket_server in daemon = %s", 
+        ((d_mode ==1)? "TRUE" : "FALSE")); 
+
+    memset(&signal_action, 0, sizeof(signal_action));
+    signal_action.sa_handler = aesdsoc_sighandler;
+    rc = sigaction(SIGINT,  &signal_action, 0);
+    if (rc < 0 ) {
+        syslog(LOG_ERR, "aesdsocket: Sigaction failed for SIGINT %s", strerror(errno));
+        goto error_1;
+    }
+    rc = sigaction(SIGTERM, &signal_action, 0);
+    if (rc < 0 ) {
+        syslog(LOG_ERR, "aesdsocket: Sigaction failed for SIGTERM %s", strerror(errno));
+        goto error_1;
+    }
+
+    soc_server = socket(AF_INET, SOCK_STREAM, 0);
+    if (soc_server < 0) {
+        rc = -1;
+        syslog(LOG_ERR, "aesdsocket: Socket creation failed %s", strerror(errno));
+        goto error_1;
+    }
+
+    rc = setsockopt(soc_server, SOL_SOCKET, SO_REUSEPORT, &cmd_option, sizeof(cmd_option));
+    if (rc < 0) {
+        syslog(LOG_ERR, "aesdsocket: API setsockopt failed %s", strerror(errno));
+        goto exit_shutdown;
+    }
+
+    aesdsoc_addr.sin_family = AF_INET;
+    aesdsoc_addr.sin_addr.s_addr = INADDR_ANY;
+    aesdsoc_addr.sin_port = htons(SOCKET_PORT);
+
+    rc = bind(soc_server, (struct sockaddr*)&aesdsoc_addr, aesdsoc_addr_len);
+    if (rc < 0) {
+        syslog(LOG_ERR, "aesdsocket: API bind failure %s", strerror(errno));
+        goto exit_shutdown;
+    }
+
+    while( exit_aesd_soc == FALSE) {
+        rc = listen(soc_server, MAX_SERVER_CONNECTION);
```

Comment:
@dwalkes : There are two while loops, the inner loop is used for completing the whole packet transfer. And then client closes the connection. Then I wait in top-level listen. What is the issue, here - it will be helpful if you could explain.

Reply:
@sujoyray-ucolorado you don't need to wait in listen, listen just sets the backlog size.  You can set it once at startup, similar to the `bind` step.

Reply:
Yes, I got that part